### PR TITLE
Added use-case testing support to upgrade testing.

### PR DESCRIPTION
**Pull request by san7ket**
Adding use cases tests, to be tested with the upgrade process.
This tends , to capture corner cases , that our traditional/functional testing tends to miss out.
This will ensure to configure entities/features/contents(CV's)/provision hosts , before upgrade do some transactions on it , aslo asserts this pre-upgarde.
After upgrade, we are able to interact with same entities/ features/hosts created pre-upgrade , perform sanity checks, add corner cases to those entities. 
issue: #532

### CODE REVIEW

File: automation_tools/satellite6/hammer.py [hunk 1]
```diff
@@ -427,3 +429,29 @@ def hammer_determine_cv_and_env_from_ak(ak_name, organization_id):
...
+    :param int prior_id: Prior lifecycle's ID
+    """
+    return hammer(
+        'lifecycle-environment create --name "{0}" '
+        '--organization-id {1} '
+        '--prior-id "{2}"'.format(
+            name, org_id, prior_id))
+
+
+def hammer_info_via_name(entity, name, org_id):
```
Comment (elyezer) [inline automation_tools/satellite6/hammer.py]: Missing docstring

File: automation_tools/satellite6/hammer.py [hunk 2]
```diff
@@ -427,3 +429,29 @@ def hammer_determine_cv_and_env_from_ak(ak_name, organization_id):
...
+
+    return hammer('"{0}" info --name "{1}" --organization-id {2}'.format(
+                     entity, name, org_id))
+
+
+def hammer_remove_cv_version(cv_version_id, cv_name, org_id):
```
Comment (elyezer) [inline automation_tools/satellite6/hammer.py]: Missing docstring

File: automation_tools/satellite6/upgrade/tests/utils.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
+import os
```
Comment (elyezer) [inline automation_tools/satellite6/upgrade/tests/utils.py]: Better to add a module docstring with more information about what type of utils we would find/place on this module.

File: automation_tools/satellite6/upgrade/tests/utils.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+
+def create_dict(dict):
+    data.update(dict)
+    print data
+    with open('{}.pkcl'.format('datastore'), 'wb') as pref:
+        pickle.dump(data, pref)
+
+
+def get_dict():
```
Comment (elyezer) [inline automation_tools/satellite6/upgrade/tests/utils.py]: This function is not used anywhere...

File: automation_tools/satellite6/upgrade/tests/utils.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+rpm1 = 'https://inecas.fedorapeople.org/fakerepos/zoo3/bear-4.1-1.noarch.rpm'
+rpm2 = 'https://inecas.fedorapeople.org/fakerepos/zoo3/camel-0.1-1.noarch.rpm'
+data = {}
+rhev_template = 'sat6-docker-upgrade-template'
+
+logger = logger()
+
+
+def create_dict(dict):
```
Comment (elyezer) [inline automation_tools/satellite6/upgrade/tests/utils.py]: Using dict name will override the dict builtin, better to use other name for the parameter

Comment (elyezer) [inline automation_tools/satellite6/upgrade/tests/utils.py]: Also reading the body of this function it seems to be storing the dict data into disk, maybe a better name and a docstring for this function would make easier to get its purpose.

File: automation_tools/satellite6/upgrade/tests/utils.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+    Environment Variables:
+
+    RHEV_SAT_HOST
+        The satellite hostname for which clients to be generated and
+        registered
+    RHEV_CLIENT_AK
+        The AK using which client will be registered to satellite
+    """
+    if int(clients_count) == 0:
```
Comment (elyezer) [inline automation_tools/satellite6/upgrade/tests/utils.py]: Better to use try/except block here:

```python
>>> int('as')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid literal for int() with base 10: 'as'
>>> try:
...   int('as')
... except ValueError:
...   print('Clients count to generate on Docker should be atleast 1 !')
... 
Clients count to generate on Docker should be atleast 1 !
```

File: automation_tools/satellite6/upgrade/tests/utils.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        logger.warning(
+            'Clients count to generate on Docker should be atleast 1 !')
+        sys.exit(1)
+    satellite_hostname = get_satellite_host()  # specify satellite hostname
+    result = {}
+    for count in range(int(clients_count)):
```
Comment (elyezer) [inline automation_tools/satellite6/upgrade/tests/utils.py]: int(clients_count) is being repeated, it looks like worth creating a variable storing the value.

File: automation_tools/satellite6/upgrade/tests/utils.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        hostname = '{0}dockerclient{1}'.format(count, client_os)
+        container_id = run(
+            'docker run -d -h {0} -e "SATHOST={1}" '
+            '-e "AK={2}" upgrade:{3}'.format(
+                hostname, satellite_hostname, ak, client_os))
+        result[hostname] = container_id
+    return result
+
+
+def get_id(entity, name, org_id):
```
Comment (elyezer) [inline automation_tools/satellite6/upgrade/tests/utils.py]: s/get_id/get_org_id?

Also it is missing docstring.

File: automation_tools/satellite6/upgrade/tests/utils.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    )[docker_vm]
+    return clients
+
+
+def docker_return_code_last(container_id):
+    result = docker_execute_command(container_id, 'echo $?', quiet=True)
+    return result
+
+
+def get_satellite_host():
```
Comment (elyezer) [inline automation_tools/satellite6/upgrade/tests/utils.py]: The `get` accepts a second parameter which is the default value if the key is not defined. This is equivalent code if one env var will available:

```python
return os.environ.get(
    'RHEV_SAT_HOST',
    os.environ.get('SATELLITE_HOSTNAME')
) 
```

If both might be available and but the first can be empty string then the code is equivalent:

```python
return os.environ.get('RHEV_SAT_HOST') or environ.get('SATELLITE_HOSTNAME')
```

File: automation_tools/satellite6/upgrade/tests/utils.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    return get_attribute_value(result, name, 'id')
+
+
+def get_latest_repo_version(cv_name):
+    cv_version_data = hammer(
+        'content-view version list --content-view {} '
+        '--organization-id 1'.format(cv_name))
+    latest_cv_ver = sorted([float(data['name'].split(
+        '{} '.format(cv_name))[1]) for data in cv_version_data]).pop()
+    cv_ver_id = get_attribute_value(cv_version_data, '{0} {1}'.format(
```
Comment (elyezer) [inline automation_tools/satellite6/upgrade/tests/utils.py]: You can return here instead of creating the single use var cv_ver_id

File: automation_tools/satellite6/upgrade/tests/utils.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+        cv_name, latest_cv_ver), 'id')
+    return cv_ver_id
+
+
+def create_vm():
```
Comment (elyezer) [inline automation_tools/satellite6/upgrade/tests/utils.py]: create_vm means I can create a VM on libvirt, RHEV? Better to add a docstring and maybe a more descriptive name.

File: automation_tools/satellite6/upgrade/tests/utils.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+        wait_till_rhev_instance_status(instance_name, 'up', 5)
+        execute(manage_daemon, 'restart', 'docker', host=docker_vm)
+    rhevm_client.disconnect()
+    time.sleep(5)
+    logger.info('Generating 1 client on RHEL6 and RHEL7 on Docker. '
+                'Please wait .....')
+    # First delete if any containers running
+    execute(
+        remove_all_docker_containers, only_running=False, host=docker_vm)
+    time.sleep(5)
```
Comment (elyezer) [inline automation_tools/satellite6/upgrade/tests/utils.py]: The below sleep can be removed and this can be update to 35

File: automation_tools/satellite6/upgrade/tests/utils.py [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+        'rhel7',
+        1,
+        ak_name,
+        host=docker_vm,
+    )[docker_vm]
+    return clients
+
+
+def docker_return_code_last(container_id):
+    result = docker_execute_command(container_id, 'echo $?', quiet=True)
```
Comment (elyezer) [inline automation_tools/satellite6/upgrade/tests/utils.py]: You can return directly without creating the single use variable.

File: automation_tools/satellite6/upgrade/tests/utils.py [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+    usecase_instance = rhev_template + 'usecase-testing'
+    delete_rhevm_instance(usecase_instance, timeout=5)
+    create_rhevm_instance(usecase_instance,
+                          rhev_template,
+                          'SAT-QE',
+                          'SAT-QE',
+                          'SAT-QE')
+
+
+def dockerize(ak_name):
```
Comment (elyezer) [inline automation_tools/satellite6/upgrade/tests/utils.py]: Not sure about what this function does, it deals with a pre created docker VM and then do some rhvm actions. I think a docstring would clarify that and maybe some more descriptive name.

File: upgrade_tests/test_scenario.py [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+    prd_name = gen_alpha()
+    repo_name = gen_alpha()
+    lc_name = gen_alpha()
+    ak_name = gen_alpha()
+    cv_name = gen_alpha()
+    docker_vm = os.environ.get('DOCKER_VM')
+    org_id = 1
+    sat_host = get_satellite_host()
+    file_path = '/var/www/html/pub/custom_repo/'
+    custom_repo = 'https://' + re.split('@', sat_host)[1] + '/pub/custom_repo/'
```
Comment (elyezer) [inline upgrade_tests/test_scenario.py]: urllib may help you building this URL

File: upgrade_tests/test_scenario.py [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+    rpm1,
+    rpm2
+)
+from fabric.api import execute, run
+from fauxfactory import gen_alpha
+from unittest2.case import TestCase
+from upgrade_tests import post_upgrade, pre_upgrade
+
+
+class Case(TestCase):
```
Comment (elyezer) [inline upgrade_tests/test_scenario.py]: Case?

File: upgrade_tests/test_scenario.py [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+
+    def tearDown(self):
+        hammer.set_hammer_config(host=self.sat_host)
+
+    @pre_upgrade
+    def test_create_product1(self):
+        try:
+            run('rm -rf {0}'.format(self.file_path))
+            run('mkdir {0}'.format(self.file_path))
+        except Exception:
```
Comment (elyezer) [inline upgrade_tests/test_scenario.py]: Exception is too generic, which exception you are expecting to be raised here?

**Comment by san7ket** [comment]
Comments addressed , though due to migration. Will raise this PR in sat-upgrade project.Once the inital PR is merged.

**Comment by san7ket** [comment]
PR Raised on new projects, resolved all earlier comments. Please revisit https://github.com/SatelliteQE/satellite6-upgrade/pull/14
